packages/vite/src/node/server/pluginContainer.ts

```diff
@@ -239,202 +265,402 @@ export async function createEnvironmentPluginContainer(
-      options: {
-        id: string
-        resolveDependencies?: boolean
-      } & Partial<PartialNull<ModuleOptions>>,
-    ): Promise<ModuleInfo> {
-      // We may not have added this to our module graph yet, so ensure it exists
-      await moduleGraph?.ensureEntryFromUrl(unwrapId(options.id))
-      // Not all options passed to this function make sense in the context of loading individual files,
-      // but we can at least update the module info properties we support
-      this._updateModuleInfo(options.id, options)
-
-      const loadResult = await container.load(options.id)
-      const code =
-        typeof loadResult === 'object' ? loadResult?.code : loadResult
-      if (code != null) {
-        await container.transform(code, options.id)
-      }
+      debugPluginResolve?.(
+        timeFrom(pluginResolveStart),
+        plugin.name,
+        prettifyUrl(id, this.environment.config.root),
+      )
 
-      const moduleInfo = this.getModuleInfo(options.id)
-      // This shouldn't happen due to calling ensureEntryFromUrl, but 1) our types can't ensure that
-      // and 2) moduleGraph may not have been provided (though in the situations where that happens,
-      // we should never have plugins calling this.load)
-      if (!moduleInfo)
-        throw Error(`Failed to load module with id ${options.id}`)
-      return moduleInfo
+      // resolveId() is hookFirst - first non-null result is returned.
+      break
     }
 
-    getModuleInfo(id: string) {
-      const module = moduleGraph?.getModuleById(id)
-      if (!module) {
-        return null
-      }
-      if (!module.info) {
-        module.info = new Proxy(
-          { id, meta: module.meta || EMPTY_OBJECT } as ModuleInfo,
-          ModuleInfoProxy,
+    if (debugResolve && rawId !== id && !rawId.startsWith(FS_PREFIX)) {
+      const key = rawId + id
+      // avoid spamming
+      if (!this._seenResolves[key]) {
+        this._seenResolves[key] = true
+        debugResolve(
+          `${timeFrom(resolveStart)} ${colors.cyan(rawId)} -> ${colors.dim(
+            id,
+          )}`,
         )
       }
-      return module.info
     }
 
-    _updateModuleInfo(id: string, { meta }: { meta?: object | null }) {
-      if (meta) {
-        const moduleInfo = this.getModuleInfo(id)
-        if (moduleInfo) {
-          moduleInfo.meta = { ...moduleInfo.meta, ...meta }
-        }
-      }
+    if (id) {
+      partial.id = isExternalUrl(id) ? id : normalizePath(id)
+      return partial as PartialResolvedId
+    } else {
+      return null
     }
+  }
 
-    _updateModuleLoadAddedImports(id: string) {
-      const module = moduleGraph?.getModuleById(id)
-      if (module) {
-        moduleNodeToLoadAddedImports.set(module, this._addedImports)
+  async load(id: string, options?: {}): Promise<LoadResult | null> {
+    const ssr = this.environment.name !== 'client'
+    options = options ? { ...options, ssr } : { ssr }
+    const ctx = new LoadPluginContext(this)
```

Comment:
IIUC `load()` and `transform()` hooks still need to create its own context everytime? Seems like Rollup does this too with its `replaceContext` param, so we still need to track `_addedImports` and call `updateModuleLoadAddedImports`.

Reply:
Yes exactly